feat(CodinGame): new presence

LighTend3r opened the pull request:
## Description 
So i have create the code to get a presence in a CodinGame site, I have added all the pages of this website, I get the code dynamically.

## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
 <summary>The home of this website</summary>

![image](https://user-images.githubusercontent.com/87587438/176157077-12035d79-f1d5-4427-8c3a-1a5290fa1a81.png)
![image](https://user-images.githubusercontent.com/87587438/176157868-9a80f059-87e4-4e57-b39b-9a7a88281330.png)

The number 15 is get with a querySelector like the rank

I tried to respect all the constraints
</details>

<details>
 <summary>Learn on this webSite</summary>

If you want learn, you can view on your profile
![image](https://user-images.githubusercontent.com/87587438/177399160-226302e8-aec3-472c-9d86-b0eb2a9d0870.png)
![image](https://user-images.githubusercontent.com/87587438/177399203-7c829239-33db-4547-aaf9-44b0be6eb82b.png)

And if you click on course, you can see the name of the course
![image](https://user-images.githubusercontent.com/87587438/177399337-0ae12765-5bd0-4c00-9966-15f6faeddda6.png)
![image](https://user-images.githubusercontent.com/87587438/177399601-51c2e9a8-ff93-449d-9a76-887bde05dc32.png)



</details>
<details>
 <summary>Not logged in</summary>

![image](https://user-images.githubusercontent.com/87587438/177400623-d431c5a5-3928-4a97-8c45-287d615d60bb.png)
![image](https://user-images.githubusercontent.com/87587438/177400711-d4e9be58-23d7-4421-894d-6138d02d5be0.png)


</details>
<details>
 <summary>some extra screen</summary>

Another exemple if you are solving a competitive challenge, you can see your rank, the name of this challenge
![image](https://user-images.githubusercontent.com/87587438/177397904-d18f984e-3ca0-4db7-8bd5-7dff8a27a85e.png)
![image](https://user-images.githubusercontent.com/87587438/177397970-0b21db97-2b63-4281-8cc3-54ead5464591.png)

If you are looking at a really difficult puzzle, you can see the difficult, the percentage of resolution (if the name it's not too long...)
![image](https://user-images.githubusercontent.com/87587438/177398141-316998af-d925-483b-b863-b8cd7aea431f.png)
![image](https://user-images.githubusercontent.com/87587438/177398164-0b92777f-6b3f-4b53-8627-dc73419cba35.png)

For exemple with this eazy chall, you can see 0% of resolution
![image](https://user-images.githubusercontent.com/87587438/177398563-d8b4e4c2-ee55-4c70-8e56-ef196ed9579c.png)

If someone is waiting a clash, you can click on the button to join the clash too
![image](https://user-images.githubusercontent.com/87587438/177398765-84f51043-d905-45ba-898c-3fa8f72243d0.png)
![image](https://user-images.githubusercontent.com/87587438/177398710-22bd9d2d-21f4-463f-9f74-ffd9694f357c.png)



</details>

LighTend3r committed:
Add files via upload

add a presence for CodinGame site

LighTend3r committed:
Merge pull request #1 from LighTend3r/LighTend3r-CodinGame-v1

add a presence for CodinGame

LighTend3r committed:
Add files via upload

LighTend3r committed:
Delete CodinGame directory

LighTend3r committed:
Add files via upload

ImRodry commented:
Please don’t edit out parts of the PR template.

LighTend3r committed:
Add files via upload

LighTend3r committed:
Fix with yarn and eslint

LighTend3r committed:
Fix with yarn prettier

LighTend3r committed:
fix a little mistake, hard -> very hard

LighTend3r commented:
Hi, I've tried everything on my side to make sure I didn't find any bug, normally, I followed all the recommendations I asked.
Thank you very much for this rigour, it allows to learn a lot of things, if you have a question, a recommendation or a problem on my code, don't hesitate to contact me

LighTend3r commented:
Looking at other people's jumpers I finally understood the commit convention, I'm sorry I didn't respect it

LighTend3r commented:
I've added some screenshots ^^

EncryptedDev commented on websites/C/CodinGame/dist/metadata.json:
```diff
@@ -0,0 +1,22 @@
+	},
+	"url": "www.codingame.com",
+	"version": "1.0.0",
+	"logo": "https://i.imgur.com/ar0zfXY.png",
```
Logo should be 512x512

EncryptedDev commented on websites/C/CodinGame/dist/metadata.json:
```diff
@@ -0,0 +1,22 @@
+	"service": "CodinGame",
+	"description": {
+		"en": "CodinGame is a web site to learn programing",
+		"fr": "CodinGame est un site pour apprendre la programmation"
```
@Kozou4ever please review this

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+			largeImageKey: "https://i.imgur.com/ar0zfXY.png",
+			startTimestamp: browsingTimestamp,
+		},
+		path = document.location.pathname;
```
It is best to leave the variable deconstructed instead of doing as you did
```suggestion
		{ pathname } = document.location;
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+		path = document.location.pathname;
+	if (path === "/") presenceData.details = "Browsing...";
+	else if (path.includes("/home")) {
+		presenceData.details = "On dashboard";
```
```suggestion
		presenceData.details = "On the dashboard";
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+			)[4].textContent
+		}`;
+	} else if (path.includes("/training")) {
+		if (path.includes("/training/easy/")) {
+			presenceData.details = "Training: Easy";
+			presenceData.state = `${
+				document.querySelector("h1[class='title']").textContent
+			} (${document.querySelector("div[class='chart-inner']").textContent}%)`;
+		} else if (path.includes("/training/medium/")) {
+			presenceData.details = "Training: Medium";
+			presenceData.state = `${
+				document.querySelector("h1[class='title']").textContent
+			} (${document.querySelector("div[class='chart-inner']").textContent}%)`;
+		} else if (path.includes("/training/hard/")) {
+			presenceData.details = "Training: Hard";
+			presenceData.state = `${
+				document.querySelector("h1[class='title']").textContent
+			} (${document.querySelector("div[class='chart-inner']").textContent}%)`;
+		} else if (path.includes("/training/expert/")) {
+			presenceData.details = "Training: Very Hard";
+			presenceData.state = `${
+				document.querySelector("h1[class='title']").textContent
+			} (${document.querySelector("div[class='chart-inner']").textContent}%)`;
```
This is repeated with the only difference of the training's difficulty, instead check if the path includes `/training/` and if it does, take the difficulty from the URL by using `document.location.pathname.split("/")[n]`, where `n` is the index value

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+				document.querySelector("h1[class='title']").textContent
+			} (${document.querySelector("div[class='chart-inner']").textContent}%)`;
+		} else presenceData.details = "Looking at Practice";
+	} else if (path.includes("/multiplayer")) {
```
Similarly with multiplayer

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+					document.querySelector("span[class='rank-suffix']").textContent
+				}/${document.querySelector("span[class='rank-total']").textContent})`;
+			}
+		} else presenceData.details = "Looking at Compete";
```
Doesn't really make sense

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+			}
+		} else presenceData.details = "Looking at Compete";
+	} else if (path.includes("/cooperate"))
+		presenceData.details = "Looking at Cooperate";
```
Ditto

EncryptedDev replied:
This still sounds weird

EncryptedDev replied:
This has not been resolved.https://github.com/PreMiD/Presences/pull/6417/files#r915031372

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+		} else presenceData.details = "Looking at Compete";
+	} else if (path.includes("/cooperate"))
+		presenceData.details = "Looking at Cooperate";
+	else if (path.includes("/learn")) presenceData.details = "Looking at Learn";
+	else if (path.includes("/events")) presenceData.details = "Looking at Events";
+	else if (path.includes("/contribute/community"))
+		presenceData.details = "Looking at Community";
+	else if (path.includes("/contribute/my-contributions"))
+		presenceData.details = "Looking at My Contributions";
+	else if (path.includes("/live-streams"))
+		presenceData.details = "Looking at Live Streams";
+	else if (path.includes("/blog")) presenceData.details = "Looking at Blog";
+	else if (path.includes("/forum")) presenceData.details = "Looking at Forum";
+	else if (path.includes("/profile"))
+		presenceData.details = "Looking at Profile";
+	else if (path.includes("/settings"))
+		presenceData.details = "Looking at Settings";
+	else if (path.includes("/about")) presenceData.details = "Looking at Abouts";
+	else if (path.includes("/careers"))
+		presenceData.details = "Looking at Careers";
+	else if (path.includes("/work")) presenceData.details = "Looking at Work";
```
Include your articles
```suggestion
	else if (path.includes("/learn")) presenceData.details = "Looking at the learn page";
	else if (path.includes("/events")) presenceData.details = "Looking at the events page";
	else if (path.includes("/contribute/community"))
		presenceData.details = "Looking at community page";
	else if (path.includes("/contribute/my-contributions"))
		presenceData.details = "Looking at their contributions";
	else if (path.includes("/live-streams"))
		presenceData.details = "Looking at live streams";
	else if (path.includes("/blog")) presenceData.details = "Looking at the blog";
	else if (path.includes("/forum")) presenceData.details = "Looking at the forum";
	else if (path.includes("/profile"))
		presenceData.details = "Looking at a profile";
	else if (path.includes("/settings"))
		presenceData.details = "Looking at their settings";
	else if (path.includes("/about")) presenceData.details = "Looking at the about page";
	else if (path.includes("/careers"))
		presenceData.details = "Looking at the careers page";
	else if (path.includes("/work")) presenceData.details = "Looking at the work page";
```
Only accept the suggestion if the suggestion is fitting for the situations

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+		path.includes("/legal/privacy-policy") &&
+		path.includes("/playgrounds")
+	)
+		presenceData.details = "Looking at Privacy Policy";
```
```suggestion
		presenceData.details = "Looking at the privacy policy";
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+	)
+		presenceData.details = "Looking at Privacy Policy";
+	else if (path.includes("/help-center"))
+		presenceData.details = "Looking at FAQ";
```
```suggestion
		presenceData.details = "Looking at the FAQ page";
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+	else if (path.includes("/start")) presenceData.details = "Not logged in";
+	else if (path.includes("/ide")) {
+		if (path.includes("/ide/puzzle/")) {
+			presenceData.details = `Solving the challenge : ${
```
```suggestion
			presenceData.details = `Solving the challenge: ${
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+							presenceData.state = `${rank}th/${
+								document.querySelector("span[class='rank-total']").textContent
+							}`;
+						} else presenceData.state = `(${score}%)`;
```
```suggestion
						} else presenceData.state = `${score}%";
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+				const score = document.querySelector(
+					"span[class*='score-value']"
+				).textContent;
+				if (score === "N/A") presenceData.state = "(0%)";
```
```suggestion
				if (score === "N/A") presenceData.state = "0%";
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+						const rank = document.querySelector(
+							"span[class='rank-value']"
+						).textContent;
+						if (rank !== "") {
```
An empty string is falsy
```suggestion
						if (!rank) {
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+							}`;
+						} else presenceData.state = `(${score}%)`;
+					} catch {
+						presenceData.state = `(${score}%)`;
```
```suggestion
						presenceData.state = `${score}%`;
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+					}
+				}
+			} catch {
+				presenceData.state = `League : ${
```
```suggestion
				presenceData.state = `League: ${
```

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+			)[
+				document.querySelectorAll("h1[class='cg-ide-title'] > span").length - 1
+			].textContent;
+			presenceData.state = `Time left : ${
+				document.querySelector("span[class*='minutes']").textContent
+			}m : ${document.querySelector("span[class*='seconds']").textContent}s`;
+		}
```
You don't have to do this, but you could use timestamps if successfully converted

LighTend3r replied:
Yes, but with this I can get exactly the time before the clash is over.

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,199 @@
+		presenceData.details = "Waiting for the Clash of Code";
+		presenceData.buttons = [
+			{
+				label: "Join the clash",
```
```suggestion
				label: "Join Clash",
```

LighTend3r committed:
fix(CodinGame): Join Clash

Co-authored-by: Rhys <64903135+EncryptedDev@users.noreply.github.com>

LighTend3r committed:
Fix(CodinGame): League:

Co-authored-by: Rhys <64903135+EncryptedDev@users.noreply.github.com>

LighTend3r committed:
Fix(CodinGame): without brackets

Co-authored-by: Rhys <64903135+EncryptedDev@users.noreply.github.com>

LighTend3r committed:
Fix(CodinGame): "On the dashboard"

Co-authored-by: Rhys <64903135+EncryptedDev@users.noreply.github.com>

LighTend3r committed:
Fix(CodinGame): "(!rank)"

Co-authored-by: Rhys <64903135+EncryptedDev@users.noreply.github.com>

itsmelouis reviewed (approved):
LGTM as regards the french description

LighTend3r committed:
Fix(CodinGame): try changing the image to 512x512

LighTend3r committed:
Fix(CodinGame): deconstructed pathname and one fix

Use variable deconstructed for pathname and remove /training for easy/medium/hard/expert

LighTend3r committed:
Merge branch 'main' into main

LighTend3r committed:
Fix(CodinGame): forgotten change of request

I didn't see that there were more requests for change, sorry.

LighTend3r committed:
Fix(CodinGame): The image size is now 512x512.

LighTend3r commented:
> Resolve conflicts as you've handled them

sorry, but I don't understand your comment

EncryptedDev commented on websites/C/CodinGame/presence.ts:
```diff
@@ -0,0 +1,205 @@
+				document.querySelector("h1[class='title']").textContent
+			} (${document.querySelector("div[class='chart-inner']").textContent}%)`;
+		} else if (pathname.includes("/expert/")) {
+			presenceData.details = "Training: Very Hard";
```
Consistency
```suggestion
			presenceData.details = "Training: Expert";
```

LighTend3r replied:
On the website, you can see Very Hard for this category, just the path name is Expert, so I think I will leave Very hard

LighTend3r commented:
If you're talking about the "else", it's when you've just clicked on the "Competition" button, and you've just https://www.codingame.com/multiplayer that link, so you're looking at the competition.

EncryptedDev commented:
> If you're talking about the "else", it's when you've just clicked on the "Competition" button, and you've just [codingame.com/multiplayer](https://www.codingame.com/multiplayer) that link, so you're looking at the competition.

In that case, wouldn't "Looking at a competition" make more grammatical sense?

LighTend3r committed:
fix(CodinGame): change "Looking at a competition"

LighTend3r commented:
Okay, it's good for me

LighTend3r commented:
![image](https://user-images.githubusercontent.com/87587438/178296546-c60f9ffc-ac29-47e4-b1b2-7cc2794c808d.png)
on my side it's good :/

LighTend3r commented:
Sorry, but I'm sure it's good, go see on my fork, in the branch main, the one linked to this PR, you can see "Looking at a competition
And I don't understand how I can fix what you see

EncryptedDev commented:
> Sorry, but I'm sure it's good, go see on my fork, in the branch main, the one linked to this PR, you can see "Looking at a competition And I don't understand how I can fix what you see
> 
> I even tried to disconnect to see if the branch changes but I still don't see a problem

I'm talking about the string "Looking at  Coorperate"

EncryptedDev commented:
> Oh, sorry, but I'm bad at English, so I didn't see that mistake. What do you think is the best translation for this ?

Looking at the corporate page

LighTend3r committed:
fix(CodinGame): "Looking at the corporate page"
